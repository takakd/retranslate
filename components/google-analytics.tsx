/**
 * Google Analytics snippet code
 */
export const GoogleAnalytics = () => {
  return (
    <template>
      {/*Global site tag (gtag.js) - Google Analytics*/}
      <script
        async
        src="https://www.googletagmanager.com/gtag/js?id=G-CPT8TKNZH7"
      ></script>
      <script
        dangerouslySetInnerHTML={{
          __html: `window.dataLayer = window.dataLayer || [];
            function gtag(){dataLayer.push(arguments);}
            gtag('js', new Date());
            gtag('config', 'G-CPT8TKNZH7');`,
        }}
      />
    </template>
  )
}
