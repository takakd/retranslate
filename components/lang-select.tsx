import React, { useEffect, useState } from 'react'

import { Dropdown, DropdownButton } from 'react-bootstrap'

import { LangType } from '../grpc/translator_pb'

/**
 * LangSelectChangeEvent is the event object of LangSelectProps.onLangChange event.
 */
export interface LangSelectChangeEvent {
  /**
   * Selected language type
   */
  lang: LangType
}

/**
 * LangSelect component props
 */
export interface LangSelectProps {
  /**
   * Language type
   */
  lang?: LangType

  /**
   * LangType change event handler.
   * @param {LangSelectChangeEvent} event event object.
   */
  onLangChange?: (event: LangSelectChangeEvent) => void

  /**
   * Language options
   */
  langOptions: {
    label: string
    value: LangType
  }[]
}

/**
 * LangSelect component
 *
 * @param {LangSelectProps} props
 */
export const LangSelect = (props: LangSelectProps) => {
  const [langLabel, setLangLabel] = useState<string>('')

  useEffect(() => {
    // Set option label by props.lang.
    setLangLabel(toLangLabel(props.lang))
  })

  /**
   * Creates the language label by LangType.
   * @param {LangType} lang language type.
   * @returns {string} language label.
   */
  const toLangLabel = (lang: LangType): string => {
    for (const option of props.langOptions) {
      if (lang === option.value) {
        return option.label
      }
    }
    return 'Language'
  }

  /**
   * Fired on clicking the language option.
   * @param {LangType} lang Selected language.
   */
  const onClick = (lang: LangType) => {
    if (!props.onLangChange) {
      return
    }
    props?.onLangChange({
      lang,
    })
  }

  return (
    <DropdownButton
      key={props.lang}
      title={langLabel}
      variant="secondary"
      renderMenuOnMount={true}
    >
      {props.langOptions.map((option) => {
        return (
          <Dropdown.Item
            as="button"
            key={option.value}
            onClick={() => onClick(option.value)}
          >
            {option.label}
          </Dropdown.Item>
        )
      })}
    </DropdownButton>
  )
}
