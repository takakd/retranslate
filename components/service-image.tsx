import React from 'react'

/**
 * Translation API service name
 */
export enum Service {
  AWS = 'aws',
  Google = 'google',
}

/**
 * ServiceImage component props
 */
export class ServiceImageProps {
  /**
   * Translation API service name
   */
  service: Service
}

/**
 * ServiceImage component
 *
 * @param {ServiceImageProps} props
 */
export const ServiceImage = (props: ServiceImageProps) => {
  if (props.service === Service.Google) {
    return <img src="/logo-google.png" alt="google" width="20" height="20" />
  } else if (props.service === Service.AWS) {
    return <img src="/logo-aws.png" alt="aws" width="28" height="20" />
  }
}
