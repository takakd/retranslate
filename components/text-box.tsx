import React, { FormEvent, useRef } from 'react'

import { CloseButton, Spinner } from 'react-bootstrap'
import { ArrowRight, Clipboard } from 'react-bootstrap-icons'

import { ServiceImage } from '../components/service-image'

/**
 * Translation service name
 */
export enum Service {
  AWS = 'aws',
  Google = 'google',
}

/**
 * TextBoxInputEvent represents an event object of TextBoxProps.onInput event.
 */
export interface TextBoxInputEvent {
  /**
   * Text in textarea
   */
  text: string
}

/**
 * Props of TextBox component
 */
export interface TextBoxProps {
  /**
   * Initial text in textarea
   */
  text: string
  /**
   * Translation service name
   */
  service?: Service
  /**
   * Input event handler.
   * @param {TextBoxInputEvent} event event object.
   */
  onInput?: (event: TextBoxInputEvent) => void
  /**
   * Clear text event handler.
   */
  onClearText?: () => void
  /**
   * Class name
   */
  className?: string
  /**
   * Class name prefix of child elements.
   */
  classNamePrefix?: string
  /**
   * Placeholder text in textarea.
   */
  placeHolder?: string

  /**
   * Show arrow on the left of textarea.
   */
  showArrow?: boolean
  /**
   * Show loading icon on the left of textarea.
   */
  loading?: boolean
}

/**
 * TextBox component.
 *
 * @param {TextBoxProps} props
 */
export const TextBox = (props: TextBoxProps) => {
  const textareaRef = useRef(null)
  const className = props?.className ? props.className : 'text-box'
  const classNamePrefix = props?.classNamePrefix
    ? props.classNamePrefix
    : 'text-box'

  /**
   * Fired on inputting texts in textarea.
   * @param {React.FormEvent} event
   */
  const onInput = (event: FormEvent) => {
    if (!props.onInput) {
      return
    }

    const target = event.target as HTMLTextAreaElement
    props.onInput({
      text: target.value,
    })
  }

  /**
   * Fired on clicking the clipboard button.
   */
  const onClickClipboard = () => {
    // Copy text in textarea to the OS clipboard.
    textareaRef.current.select()
    document.execCommand('copy')
  }

  return (
    <div className={`d-flex flex-row align-items-center w-100 ${className}`}>
      {
        // Show arrow icon.
        props.showArrow && !props.loading && (
          <div className="icon">
            <ArrowRight size={20} />
          </div>
        )
      }
      {
        // Show loading icon when loading.
        props.loading && (
          <div className="icon">
            <Spinner animation="border" size="sm" />
          </div>
        )
      }
      <div className={`d-flex flex-column ${classNamePrefix}-text-container`}>
        {
          <div className={`${className}-header`}>
            {
              // Show service icon
              props.service && <ServiceImage service={props.service} />
            }
          </div>
        }
        <div
          className={`position-relative w-100 ${classNamePrefix}-textarea-container`}
        >
          <textarea
            ref={textareaRef}
            className={`h-100 ${classNamePrefix}-textarea`}
            placeholder={props?.placeHolder}
            defaultValue={props.text}
            onInput={(event) => onInput(event)}
          />
          {
            // Show erase button when text is entered and the event handler exists.
            props.text !== '' && props?.onClearText && (
              <CloseButton
                className={`position-absolute close-button p-1 ${classNamePrefix}-close-button`}
                onClick={props?.onClearText}
              />
            )
          }
          {
            // Show clipboard button when text is entered.
            props.text !== '' && (
              <Clipboard
                className={`position-absolute ${classNamePrefix}-clipboard-button`}
                onClick={onClickClipboard}
              />
            )
          }
        </div>
      </div>

      <style jsx>{`
        .${className}-header {
          min-height: 30px;
        }

        .${classNamePrefix}-textarea {
          width: 100%;
          resize: none;
          padding: 0.5rem 0.5rem 36px 0.5rem;
        }
        .${classNamePrefix}-text-container,
          .${classNamePrefix}-textarea-container {
          flex: 1;
        }
        .icon {
          min-width: 30px;
          text-align: center;
        }
        ::placeholder {
          color: #aaa;
          opacity: 1;
        }
      `}</style>
      <style jsx global>{`
        .${classNamePrefix}-close-button {
          top: 1px;
          right: 6px;
        }
        .${classNamePrefix}-clipboard-button {
          cursor: pointer;
          bottom: 10px;
          right: 10px;
        }
      `}</style>
    </div>
  )
}
