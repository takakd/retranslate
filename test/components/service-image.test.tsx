import React from 'react'

import { Service, ServiceImage } from '../../components/service-image'
import { render } from '../testUtils'

describe('ServiceImage', () => {
  it('matches google snapshot', async () => {
    const { asFragment } = render(<ServiceImage service={Service.AWS} />)

    expect(asFragment()).toMatchSnapshot()
  })

  it('matches aws snapshot', async () => {
    const { asFragment } = render(<ServiceImage service={Service.AWS} />)

    expect(asFragment()).toMatchSnapshot()
  })
})
