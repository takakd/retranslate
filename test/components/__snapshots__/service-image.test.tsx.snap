// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ServiceImage matches aws snapshot 1`] = `
<DocumentFragment>
  <img
    alt="aws"
    height="20"
    src="/logo-aws.png"
    width="28"
  />
</DocumentFragment>
`;

exports[`ServiceImage matches google snapshot 1`] = `
<DocumentFragment>
  <img
    alt="aws"
    height="20"
    src="/logo-aws.png"
    width="28"
  />
</DocumentFragment>
`;
